### ISSUE TITLE
Detect .zip and .tar.gz by file extension, add --single-file option

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/butler#58]
This is my proposed solution for #51

It seems silly to force people to create directories with only one file in them if they just want to upload one file.

At the same time, I don't feel comfortable adding metadata such as "this should actually end up being a compressed zip file with a custom file extension" — wharf works on folders/file hierarchies.

It looks like diffing .jars (also zip-based) doesn't yield too bad results? http://myrealitycoding.github.io/2016/05/31/deploy-to-itch.html — if someone has subsequent builds of a .love game, I'd be curious to try diffing them.

**Comment by slowriot** [comment]
+1 for --single-file option; in fact I'd suggest detecting when there is only a single file being uploaded, and serving that to the user automatically, rather than serving a zip.

If storing files in zips is preferred internally for whatever reason, the file could perhaps be extracted server-side at the point when the user chooses to download it; this would make for a much smoother user experience than having to unzip a file containing a single file.  That situation is even worse for users on MacOS, where they have to unzip the file to extract a disk image DMG file, which they then have to mount to extract the app anyway as per the usual workflow.  Anything that annoys users less is to be preferred, I think.

**Comment by fasterthanlime** [comment]
I definitely agree with the UX argument - and if it was trivial, I'd have already done it.

The thing about .dmg, installer .exes, .apk files, etc. - is that they're all archives anyway. Using them pretty much kills the advantages of using a diffing/patching system.

For Android, there's no choice - it's .apk or nothing. But then again, itch.io isn't really targeted towards Android users (there's the platform tag, but that's about it), so that doesn't make it a priority.

For Mac, there *is* a choice. .dmg is the nicest UX for manual, one-time installs of games - but having the itch.io backend be aware of the files that actually make up the app means having all the advantages of diffing/patching available for macOS users too, in particular:

  * Extract-while-download installs via app (no extra steps, no image mounting, etc., on-the-fly verification)
  * Integrity verification (already in app)
  * Patching the game in-place (instead of redownloading the game entirely & unpacking it on top of the old version)
  * Smaller downloads (patches are much smaller than, well, redownloading the full game)

If the goal was just to improve UX for mac users that *don't* use the itch.io app, I'd just add .dmg generation to the itch.io backend, that seems like a much better solution.

TL;DR **butler needs to know what's inside your game to work efficiently**. It only supports "uploading .zips" because it knows how to extract them beforehand. Adding single-file support makes it a "dumb" uploader and makes the experience worse for app users.

The fact that zip is used for canonical storage server-side is just an implementation detail - we could definitely serve something else to non-app users: installers on Windows, .dmgs on macOS, and, sure, the naked .apk file for Android users. It's just a matter of where we spend our limited resources, and at the moment, butler covers the main usecase pretty well, so we're spending time/effort elsewhere!

**Comment by slowriot** [comment]
I agree with you that installers are antithetical to the concept of a diff/patch based system.  However, using the butler on itchio for us, as I suspect many other developers, simply a way of automating the process of uploading builds - the resultant product we can offer our customers shouldn't depend on the back-end tool, or be limited by it.  Just because an installer isn't optimal for the system, doesn't mean that same installer in a zip is a reasonable alternative.

As it happens, our own builds are entirely monolithic binaries - no installers, no scripts, no filesystems (except on macos), all resources compiled into the executable itself - so when we had manually uploaded files, the user could simply download the windows exe or elf binary, and double click to run it right away.  With mandatory zip files, we lose that convenience, and it's another hoop for the user to jump through.

With mac I it's rather different, as you say; dmg is the canonical way to distribute, and the way frameworks and icons work means that it's impossible to avoid.  But the current situation of a dmg inside a zip, makes the small inconvenience mac users are already used to into a much greater inconvenience.

The thing is, optimising everything for the itch app is well and good, but it's really not helpful if it breaks the user workflow for players using the website - and anyone we send to download our games gets links to the web storefront.  Those files should be first class citizens.

We also really don't want to have a separate build process for each platform just for itch, when we can upload monolithic files or our own-rolled dmgs everywhere else we distribute them - we already have the dmg built by our own processes, with the flags we need and optimised for the correct size for what we're storing,  - so server-side DMG generation wouldn't be much use to devs like us, although I can see it being helpful to those who don't have their own integrated build system already.

To reply to your TL;DR though - I understand that butler **benefits** from understanding the content of the game's structure to optimise the app user experience - however, developers like us have **nothing** in the app structure that is externally visible, simply a monolithic binary, and I think the web user experience (the vast majority of our users) shouldn't be handicapped in the vain hope of saving a couple of megabytes of bandwidth for the app users. 

The zip thing might make sense for a lot of developers, but it doesn't make sense for us -which is why I'm in favour of a flag or automatic single-file detection, so you can offer the best of both worlds.

**Comment by fasterthanlime** [comment]
> Just because an installer isn't optimal for the system, doesn't mean that same installer in a zip is a reasonable alternative.

I agree! Installers are generally bad because:

  * Only few of them are well-behaved:
    * Some require administrator rights (elevation) for install
    * Some don't allow specifying an install folder (they force install to `C:\Program File` or the x86 variant)
  * They result in non-portable installations (some files end up outside the install folder)
  * No diff/patching is possible
  * They usually do compression, but often with outdated algorithms that are easily outperformed with lzma, brotli, or zstd

However, monolithic binaries and .dmg(s) aren't much better.

### Windows

For monolithic binaries, there's two cases: either the bundled assets are compressed (using gzip, for example), and that means players need to pay the cost of decompression on every game launch - this isn't ideal.

Or, they don't do compression, and that means serving the monolithic binary directly to web users results in a longer download & more bandwidth used. The "extra hoop" of the .zip here isn't useless! It saves quite a few bytes (and you pay for decompression only once). Besides, if users want to organize their games properly, they still have to move the binary from their Downloads folder to a game folder somewhere (possibly on a different disk) - which isn't much less trouble than extracting an archive.

There is another problem with monolithic binaries: they hide the real executable as if it were data. This gets in the way of mechanisms like automatic dependency detection. Recent versions of itch shipped the [Prerequisites feature](https://itch.io/docs/itch/integrating/manifest.html) which allows installing required dependencies on a game's first launch. However, many developers omit manifests from their games, and most required frameworks can be automatically detected, provided the app can probe the *actual* executable of the game.

### Linux

For Linux, monolithic binaries can be a source of problems too - I haven't dug into how you're doing it exactly (but feel free to post a link), but the only reliable way to ship a game on Linux in a cross-distro way is to ship the libraries you're using as well, both in 32-bit and 64-bit flavor, along with a launcher script that picks the right architecture and adjusts the `LD_LIBRARY_PATH` so that they're found by the operating system (see [Linux builds](https://itch.io/docs/itch/integrating/platforms/linux.html) in the itch guide).

One can imagine a monolithic binary doing all of the above (especially if it's a set of files disguised as a self-extracting archive, maybe as a shell script with a gzip payload appended), but it's unnecessarily complicated and, again, doesn't help UX much.

> so when we had manually uploaded files, the user could simply download the windows exe or elf binary, and double click to run it right away.

On Linux, that's not true. Downloading a monolithic binary also requires an extra step as well: changing its permissions so it has the "executable bit". Whether it's done graphically, via the desktop environment's file manager, or on the command-line, with `chmod +x`, it's a step you don't need if you ship an archive.

Some folks will argue that .zip isn't an adequate archive format, saying it "doesn't preserve permissions", but that hasn't been true for a long time. Standard zips include file permissions, and compliant unarchivers preserve them when extracting. What they're seeing usually is the result of using an old or non-compliant archiving program when creating the .zip in the first place.

Another common problem when distributing Linux builds is doing it from Windows, which doesn't have the concept of "executable bit" - this results in archives where the binary lacks the executable bit. Fortunately, itch.io solves the problem on two levels:

  * butler detects Linux & macOS binaries and fixes permissions on upload
  * the backend also detects Linux & macOS binaries and ensures proper permissions when generating the .zip archive

### macOS

DMG files are a bit different. They do compression & integrity checks, which the .zip format does as well. They allow shipping signed binaries - which you can do in a .zip archive as well. The only clear advantage of DMG files is that they appear as a "disk" for the user, and you can customize the looks of the Finder window that pops up when mounting them. That's it. A downside is that users routinely forget to eject volumes, and are puzzled when dragging the DMG file to trash doesn't work.

You mention DMG as being "the canonical way to distribute apps on mac", but that's not entirely true. The canonical way is an app bundle, which is a folder, and it's not uncommon for vendors to choose to ship it compressed as a .zip rather than a .dmg.

If you don't have an Apple code signing certificate, using DMG won't save you - most users will still be unable to open & run the game, whereas they'd have no issue with the itch.io app, which works around this issue.

### Summary

In my view,

  * The friendliest UX for all platforms is using the app. Telling buyers to "just use the app" is the sane thing to do, as Microsoft & Apple keep locking up their platform, and Linux was never particularly friendly to ship for in the first place.
  * Some users will insist on *not* using the app, regardless of the advantages it provides, and we shouldn't force them. However, these users are generally tech-savvy enough to extract an archive properly, whether it is on Windows, macOS, or Linux.

Quoting your post:

> the resultant product we can offer our customers shouldn't depend on the back-end tool, or be limited by it.

I'd argue that using monolithic binaries for Windows & Linux is actually a result of being limited by *other platforms*, which don't provide an app or diff/patching in the first place.

I hope that, given the arguments above, you can see why single file support isn't a priority for me at the moment. It's a special case, providing a worse UX in many situations, and adding it would set a precedent for other developers - if there is a `--single-file` option, why bother shipping your game in the best way possible when you can just upload whatever you already have and call it a day?

This would just result in a degraded experience for users, which I really want to avoid. On the other hand, I would also really like to avoid this response sounding adversarial and discouraging developers from publishing their games on itch.io as a result. I just want to emphasize that the way things are set up right now are not a product of half-assery or laziness, but the result of carefully considered design decisions.

If the single file option is ever implemented in the future, it should probably come with a disclaimer explaining the many downsides of choosing this particular delivery format.

**Comment by slowriot** [comment]
This is a very interesting discussion, that I'd love to get into in more detail, but I'm worried about going too far off-topic here in the comments for an issue, so I'm going to try to reply in brief - I suspect it'll be a long one anyway, though, as there are a lot of issues to cover.  I'd be very happy to discuss some of these issues elsewhere though.

Re. monolithic binaries:  I disagree that they're not much better; it all depends on how intelligent the implementation is.

In our case there is no "real binary" that's hidden, as you say; there isn't some on-demand compressing wrapper or something like that.  In our case, the binary really is the game itself, and all resources it requires are compiled in at the object level, rather than accessed as filesystem elements.  This is true on windows, linux and even macos as well (although the latter just requires external metadata and icon files to be presented correctly to the user as a desktop item). 

The compiled-in assets are inevitably compressed formats anyway, optimised for the type of data - music is stored in ogg vorbis, images in png or jpeg, and so forth.  The user will always be paying at runtime to decompress those sorts of assets, for all games.  There are only trivial gains to be had from re-compressing the resulting binary, and that's mostly obtained from compressing the text strings in the executable itself - for instance, our release AdvertCity occupies 146MB on disk on Windows, and the itch zipped download weighs 142MB.  That four megabytes represents a much longer time spent by the user decompressing the file and messing about with a clumsy zip file, than they would spend simply downloading the full binary.

As for how we achieve this on multiple platforms, it's a long story, but in summary:
* For all platforms, we have developed a set of build tools that allow us to treat binary assets as compileable data, and produce linkable objects; we then use some cross-platform preprocessor glue to be able to access those directly as char arrays or whatever is the appropriate format.  This has the huge advantages of not caring about the underlying filesystem, not having to waste time at runtime checking for valid paths, files all being present, permissions being correct, and loading and streaming those files - in fact the operating system's paging mechanism ensures that the data is loaded into memory just when it needs to be, providing a much more optimal loading experience than by pre-loading files manually.  We talk about this briefly in our (currently somewhat neglected) devblog: http://voxelstorm.com/devblog/sphereface-dev-tools/.
* On Windows, we build with MinGW-W64 and link all libraries statically - no dll hell.  The only exception is steamapi and openvr; in the former case we require that when distributing on Steam only, and in the latter case we have a soft loader which fails gracefully and continues startup if the dynamic library isn't present on the system.  We currently provide just 32bit builds for Windows, since those run everywhere.
* On Linux, we link almost all libraries statically, with the exception of glibc6.  To ensure compatibility, we have built a set of helper tools to inspect symbol versions in our output binaries, and check that the maximum version of glibc6 required is below our target version - currently 2.19, which is what's available in Debian Stable and all derivatives.  The tool is smart enough to automatically suggest code to insert to request a lower symbol version - for example, to require an older version of quick_exit, it would suggest we prepend `__asm__(".symver quick_exit,quick_exit@GLIBC_2.10");` to the quick_exit call.  We also disable PIE, for both compatibility and performance reasons.  These tricks allow us to provide a completely monolithic binary that's portable to all reasonably up to date linux systems.  We provide separate 64bit and 32bit builds on Linux.
* On MacOS, we also build with GCC and link against glibc.  Most libraries are static again, and required frameworks are mainly system ones - with the exception of the C library itself.  We have a toolchain to inspect the binary for external dependencies under /usr/local (such as glibc), and to automatically package those in the app framework it assembles at build time, and re-target that library using the macos `install_name_tool -change` fudger so all search paths are local to the app; this isn't precisely a monolithic binary anymore, but the resultant program is still entirely self-contained within its .app structure.  This is then packaged in a dmg by tools that create the correctly sized disk image with enough leeway calculated in for additional metadata and other bits and pieces, and compress and read-only-ify this using the right set of arcane flags and options.  Having developed these tools over a long period with much learning from mistakes, I wouldn't want to trust a third party dmg-generator to get this right for us.

When I said "dmg is the canonical way to distribute", I meant the standard .app structure, packaged in a dmg.  This is basically universal for mac software.  I personally have no love at all for the dmg format, but it is how the vast majority of macos programs are distributed (along with a friendly "drag to applications" graphical whatsit that I'm pretty sure you can't achieve with zip alone, and which the itch app itself uses).  In this case we use that standard simply following the principle of least surprise - and this is especially important here, as many of our mac customers tend to be less technical than our users on the other two platforms.  The drag-from-dmg-to-applications paradigm is one that all mac users are familiar with.

I realise this isn't necessarily a typical scenario, but that's why I'm petitioning to give competent content authors the flexibility to use the upload tool as we wish, and deliver content to our users in the format that we choose.  Besides, it makes no sense to **not** offer this option with the butler, since the current workaround is simply to have someone wasting half an hour a day dragging and dropping in the website to upload new builds for multiple games by hand - and the user experience would be the same.


With regards to the app - yes, sure that's the most friendly option - once they already have it.  That allows automatic updates and generally makes our life easier, too.  But we're not going to tell our users to go download a third party app just to play our one game - whatever it is.  We won't tell them to downlaod itch any more than we'd tell them to download steam to play our games, or desura from when their app was a thing.  And when we bundle our games on USB drives and so on, we certainly aren't going to redistribute third party apps or require an internet download just to support those - and we aren't going to run totally separate parallel build systems just to service yet another app that wants to do things differently.  I disagree that tech savvy users are the only ones who wouldn't want to have to participate in an entire marketplace ecosystem just to play one game.  I also realise there's rather a conflict of interest here, because now rather than bringing people to play our games on our storefront, we're taking them to a larger marketplace and encouraging them to browse everyone else's games - why should we bring customers for itch when we can simply offer the downloads ourselves on our own site?

If you don't implement a single file option, the alternatives open to us currently are:
a) Continue using our uploader tool that works with the butler and uploads builds at the same time as it uploads to steam, which works fine for steam and itch the app, but results in us shipping binaries-in-zips to web users.
b) Revert to manually uploading each binary to the itch website as we did prior to uploading to the butler; this wastes precious development time,  does no favours for the itch app users, but provides the optimal web itch user experience.  Currently the latter are our priority, but it doesn't look like it's worth the time to do this manually in terms of cost/benefit.
c) Switch to self-hosting our binaries, and adapt our upload tool accordingly; provide those binaries by link from the itch website.  Unless there was a way to provide parallel upload channels for the itch app only, which would be hidden from the web users, we would need to discontinue binary uploads to itch altogether - we see very little traffic through the app, so this would not be something that would be seriously missed at present.

None of these is ideal;  Of course we'd like to make our content available on the itch app as well as the other locations, but right now it just isn't popular enough to justify the penalty you seem to be intentionally imposing on web users.  So we'll probably just end up continuing with option a, as that's the least dramatic - but under protest, as it interferes with the web user experience needlessly.  It just doesn't make sense to me to say "you can download this binary directly, if someone has taken the trouble to spend half an hour a day every day dragging and dropping - but our own automated tools provide these clumsy zips instead".  It's nothing other than shooting yourself in the foot, and it makes us look less professional, as if it were a developer choice to provide this dmg-in-a-zip option.

The alternative IMO is simple, and really wouldn't have the negative impact you seem so sure it would.  Forget about a `--single-file` option; simply automate it such, that if someone *is* uploading a directory containing only a single file using the butler, nothing changes - it's all delivered to the itch app as standard, it's diffed in the same ways - but when a user chooses to download the file from the webpage, the inner file is served, and not the outer zip.  That's all that you'd need to do to make us happy and our users happy and keep the butler system optimal.  This would also solve several other issues on this tracker, such as the .love file issue, and distributing .jars and other archives.

If you're seriously concerned about the extra bandwidth of downloading the non-zipped larger versions of single-file archives, just have the server check the decompressed size versus the compressed size, prior to offering the download to the user - if it's more than, say, 50% or 25% larger, just automatically offer the zip instead, otherwise decompress it server-side and serve the inner file.

**Comment by fasterthanlime** [comment]
Thanks for the long-form response! It's rare to see this level of detail & research.

> Forget about a --single-file option; simply automate it such, that if someone is uploading a directory containing only a single file using the butler, nothing changes - it's all delivered to the itch app as standard, it's diffed in the same ways - but when a user chooses to download the file from the webpage, the inner file is served, and not the outer zip.

I'll look into this in the near future. Or maybe @granpc could?

**Comment by slowriot** [comment]
Thank you :)

**Comment by fasterthanlime** [comment]
Just a quick belated update: @GranPC wrote the code for that feature, it should be deployed soon if I'm not mistaken @leafo ?

**Comment by slowriot** [comment]
Still waiting... we're about to have a major release, so it would be nice to streamline the players' download experience in anticipation of that.

**Comment by fasterthanlime** [comment]
Everything's ok on my side, waiting up on @leafo to deploy his, I'll ping you when it's done @slowriot :star2:

**Comment by fasterthanlime** [comment]
@slowriot everything has been deployed!

Pushing a folder or .zip with a single file in it will now result in that single file being served to web users.

App works as usual.

:star2: :star2: :star2:

**Comment by Deozaan** [comment]
What does this mean for .apk files?

**Comment by fasterthanlime** [comment]
> What does this mean for .apk files?
> 
> I should push a folder containing only the .apk and people will be able to download the .apk without any modifications?

Correct!

**Comment by slowriot** [comment]
@fasterthanlime and @leafo - nice, thank you very much!  Confirmed working.  I had to re-upload all packages to replace the zip version with the standalone files, but now I've done that, all is good.
